### ISSUE TITLE
[FEATURE] Support SPARQL pre-bound variables using initBindings?

### ISSUE THREAD

**Issue by ashleysommer** [oxigraph/oxigraph#1123]
Many SPARQL service implementations provide a mechanism to "pre-bind" a SPARQL variable with a given value.
This is often called "pre-bound variables" or "variable pre-binding", and it is part of the RDF4J query API. In RDFLib, this concept is called "initBindings".

This is different than appending a `BIND` or `VALUES` clause to the SPARQL query, the difference is that the pre-bound values are known before query execution starts, and they are present inside the deepest nested sub-SELECT statements (because SPARQL queries are executed inside-out).

On the technical side, pre-binding variables adds a row to the solution table with the given variable values, before the query execution starts.

To illustrate, take this example query:
```
SELECT ?this ?v WHERE {
    {
        SELECT ?this ?param WHERE {
            {
                SELECT ?this ?class WHERE {
                    ex:MainInstance ex:hasClasses ?class . 
                }
            }
            ?this a ?class .
            ?class ex:useParam ?param .
        }
    }
    ?this ?param ?v .
}
```
When run with `initBindings={"this": "https://example.com/myValue"}` will be evaluated like this:
```
SELECT ?this ?v WHERE {
    {
        SELECT ?this ?param WHERE {
            {
                SELECT ?this ?class WHERE {
                    BIND (<https://example.com/myValue> as ?this)
                    ex:MainInstance ex:hasClasses ?class . 
                }
            }
            ?this a ?class .
            ?class ex:useParam ?param .
        }
    }
    ?this ?param ?v .
}
```

**Motivation**
I am the author of PySHACL. I'm trying to add Oxigraph as a first-class supported store implementation for use with PySHACL.
The SHACL Spec relies on pre-bound variables when executing SPARQL-constraints. It requires the use of a store that supports pre-bound variables. https://www.w3.org/TR/shacl/#sparql-constraints-prebound

RDFLib does have a workaround, when a SPARQL service does not support initBindings, it adds a VALUES clause to the query instead, but that is an imperfect solution, as described here: https://github.com/oxigraph/oxrdflib/issues/11

I'm facing that issue ([oxrdflib #11](https://github.com/oxigraph/oxrdflib/issues/11)), and another issue where the value for a pre-bound variable is a BNode. RDFLib injects a `VALUES` clause like this:
```
VALUES ( ?this ) { (_:d5b01f057cd7e66262672e4ec17e681f) }
```
Being a BNode ID, oxigraph does not like that.
It throws this error:
```error
SyntaxError: error at 55:26: expected UNDEF
```
Even though there is no UNDEF, oxigraph treats the BNode ID as unbound.

My thought is native support for pre-bound variables in oxigraph would fix ([oxrdflib #11](https://github.com/oxigraph/oxrdflib/issues/11)) and this BNode issue (if oxigraph supports externally supplied BNode identifiers).

**Comment by ashleysommer** [comment]
Note, the SHACL spec does list a set of limitations that should be imposed on queries that use supplied pre-bound variables.

https://www.w3.org/TR/shacl/#syntax-rule-pre-binding-limitations

* SPARQL queries must not contain a `MINUS` clause
* SPARQL queries must not contain a federated query (`SERVICE`)
* SPARQL queries must not contain a `VALUES` clause
* SPARQL queries must not use the syntax form ​​`AS ?var` for any pre-bound variable ?var
* [Subqueries](https://www.w3.org/TR/sparql11-query/#subqueries) must return all pre-bound variables

I think most of them make sense, given the mechanism of operation.

**Comment by Tpt** [comment]
Thank you @afs!

I am planning to allow some "pre-bind"/"prepared" query feature based on [refined SPARQL exist substitution operation](https://github.com/w3c/sparql-dev/blob/main/SEP/SEP-0007/sep-0007.md). Thanks @afs for defining it.

Right now you can hack using `LATERAL` join:

If you have a query `SELECT ?a ?b ?c WHERE { W }` and want to pre-bind `?b` and `?d` you can rewrite the query:
```sparql
SELECT ?a ?b ?c WHERE {
   VALUES (?b ?d) { (b_value, d_value) }
   LATERAL { W }
}
```
and get what you want

**Comment by ashleysommer** [comment]
@afs Thanks for the background info an the links. I literally only just learned about the `LATERAL` join yesterday, and now I see it popping up here. I have a feeling I'm going to find it very handy when various SPARQL services start implementing it.

@Tpt 
I'm glad to know you're already planning a pre-bind/"prepared-query" feature. It is one of the features of the RDF4J API, so is required to support the implementation in https://github.com/oxigraph/oxigraph/issues/1047.

I will try to get the PySHACL use-case working using your `LATERAL` workaround, however I think this still doesn't help with the case where the value is a Blank Node. Maybe I could try something like this:
```
SELECT ?this ?v WHERE {
   BIND (BNODE("fb3e7b46722fc12ffc8089beacf346d1") as ?this)
   LATERAL { W }
}
```

**Comment by Tpt** [comment]
>  I think this still doesn't help with the case where the value is a Blank Node. Maybe I could try something like this:

Indeed. Missed that. Yes, your workaround will work in Oxigraph even if it is not mandated by [the SPARQL specification](https://www.w3.org/TR/sparql11-query/#func-bnode).

**Comment by Tpt** [comment]
#1135 allows to set variables to be substituted using the SEP-0007 approach. This has some tricky side effects (it is not allowed to substitute queries that are not in the `SELECT` part of the query but has the advantage of being well defined.

I thought about lexical replacement (replace all occurrences of the variable in the query string) but it would have weird behavior when replacing with a blank node (I guess the user want to use the actual blank node identifier and not get a variable replaced by an other variable and deal with blank nodes syntactic restrictions).

**Comment by Tpt** [comment]
The `Store.query` method now allows to substitute variables with RDF terms following SEP 0007 approach. [Documentation](https://pyoxigraph.readthedocs.io/en/stable/store.html#pyoxigraph.Store.query)

**Comment by Tpt** [comment]
Done in v0.4.6
